packages/shopping/public/shoppy.js

```diff
@@ -0,0 +1,35 @@
+const productTemplate = `
+<div id="card-#ID#" class="card" style="width: 18rem;">
+  <img src="#IMAGE#" class="card-img-top product-img" alt="#NAME#">
+  <div class="card-body">
+    <h4 class="card-title">#NAME#</h4>
+    <h5>$#PRICE#</h5>
+    <p class="card-text description">#DESCRIPTION#</p>
+    <div class="action-buttons">
+      <a href="javascript:addToCart('#ID#', '#NAME#', '#PRICE#', #UNFORMATTED-PRICE#, '#IMAGE#')" class="btn btn-primary cart-action-button">Add to Cart</a>
+      <a href="/product.html?id=#ID#" class="btn btn-primary">Details</a>
+    </div>
+  </div>
+</div>
+`;
+
+$(function () {
+  const skip = (getCurrentPageNumber() - 1) * 4;
+  api.getProducts(function (products) {
+    if (products) {
+      products.forEach(product => {
+        const productHtml = productTemplate.replace(/#NAME#/ig, product.name)
+                            .replace(/#PRICE#/ig, new Intl.NumberFormat('en-US', { maximumSignificantDigits: 3 }).format(product.price))
+                            .replace(/#UNFORMATTED-PRICE#/g, product.price)
+                            .replace(/#IMAGE#/ig, product.image)
+                            .replace(/#DESCRIPTION#/ig, product.description)
+                            .replace(/#DETAILS#/ig, product.details)
+                            .replace(/#ID#/ig, product.productId);
```

Comment:
Have you considered using ES6 template literals? (See [Browser compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Browser_compatibility).)

```ts
const productHtml = `<div id="card-${product.productId}" class="card" style="width: 18rem;">
   <img src="${product.image}" class="card-img-top product-img" alt="${product.name}">
   (etc.)
 </div>
`;
```

More importantly, please remember to HTML-escape values when building HTML strings manually, we don't want our example application to be vulnerable to [cross-site scripting](https://www.owasp.org/index.php/Testing_for_Cross_site_scripting). Consider what happens when e.g. `product.description` contains the string `<script>window.alert("you have been hacked!");</script>`.

I think the easiest option is to use Handlebars for templating, see https://handlebarsjs.com. You can add the library to your front-page via a public CDN and it will handle tricky parts of HTML escaping for you.

If we had a webpack/parcel/browserify build pipeline for front-end assets in place, then I would recommend to get rid of jQuery and use something like https://www.npmjs.com/package/nanohtml

Reply:
All great feedback!

For this PR I want to keep things simple and focus on making things works. There will be another PR after this which will handle authorization. In the third PR, let's optimize UI implementation details, tooling, and testing (https://webdriver.io/).

Reply:
Fair enough. Please create a list of follow-up tasks so that we don't accidentally forget to address parts of my feedback provided in this pull request.

To be clear, a list of follow-up tasks does not have to be a series of GitHub issues, a checklist in a single GH issue may be a better solution.

> https://webdriver.io/

I am not very familiar with front-end testing tools, but I heard a lot of good feedback on
https://www.cypress.io. You may want to give it a try.

Reply:
Sure!

Reply:
>  In the third PR, let's optimize UI implementation details, tooling, and testing

Well, I don't consider security vulnerabilities as "UI implementation details", but I won't block landing of this pull request on that 🤷‍♂

Reply:
@bajtos follow up tasks - https://github.com/strongloop/loopback4-example-shopping/issues/478.

Added DOMPurity for sanitizing input.